### PR TITLE
3D hotspots info for the annotation column

### PR DESCRIPTION
**Pull request by onursumer**
# What? Why?
Adds 3D hotspots information into the annotation column of the mutation table.

# Checks
- [ ] Runs on Heroku.
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)
- [ ] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to
  hotfix.

### CODE REVIEW

**Comment by onursumer** [comment]
We still need to add 3D hotspots publication information into the tooltip.

File: portal/src/main/webapp/WEB-INF/jsp/mutation_details.jsp [hunk 1]
```diff
@@ -76,6 +78,23 @@
         var sampleArray = window.QuerySession.getSampleIds();
         OncoKB.setUrl('<%=oncokbUrl%>');
         var mutationProxy = DataProxyFactory.getDefaultMutationDataProxy();
+        var annotationCol = null;
+
+        if(OncoKB.getAccess()) {
+            var oncokbInstanceManager = new OncoKB.addInstanceManager();
```
Comment (onursumer) [inline portal/src/main/webapp/WEB-INF/jsp/mutation_details.jsp]: @zhx828 Should we also move the initialization of OncoKB instance manager into the AnnotatorColumn class as well? Are we using OncoKB anywhere else?

**Comment by onursumer** [comment]
annotation_3d_hotspots

**Comment by onursumer** [comment]
Just realized that `OncoKBConnector` calls the `cbio.util.getHotSpotDesc()` function which is modified in this PR.

@zhx828 Is it okay for now to ignore 3D hotspots for `OncoKBConnector`?

File: portal/src/main/webapp/WEB-INF/jsp/mutation_details.jsp (hunk 1 above)
Comment (zhx828) [inline portal/src/main/webapp/WEB-INF/jsp/mutation_details.jsp]: Totally. I'm using this in patient view as well. Any plan to move to patient view?

**Comment by zhx828** [comment]
@onursumer yes. And I will move hotspot related info out from OncoKB soon.

**Comment by onursumer** [comment]
@jjgao Updated the code to use proxy instead of direct AJAX call to 3dhotspots.org. HTTPS instances should work fine now.

**Comment by onursumer** [comment]
Seems like sorting of the annotation column is handled within the `OncoKBConnector`. Sort/compare functions need to be modified to make sorting work for 3D hotspots.

**Comment by jjgao** [comment]
@onursumer you mean sorting of MyCancerGenome and Hotspot were also coded in `OncoKBConnector `?

**Comment by onursumer** [comment]
@jjgao seems like they were coded there: https://github.com/cBioPortal/cbioportal/blob/a8de0f1ac8a18f72356afbf03958a268a492f8a6/portal/src/main/webapp/js/src/oncokb/OncoKBConnector.js#L1315

**Comment by onursumer** [comment]
@jjgao it seems like whenever there is an OncoKB related icon in the annotation column, hotspot icons are not taken into account when sorting. I don't see any logic like: sort "by oncokb sensitivity" and then "by oncokb resistance" and then "by mycancergenome" and then "by cancerhotspots". It only sorts by one of those categories.

@zhx828 please correct me if I am wrong.

**Comment by zhx828** [comment]
@onursumer it does consider different annotations on sorting, at the same line of the code you pointed above, it is a for loop, it calculates the sorting based on the priority, the lower index, the higher priority. If the comparison is 0 at line 1330-1332
```
if (result !== 0) {
    return result;
 }
```
then it will compare next annotation column. You just need to add 3d hotspot into categories array, and add your own comparison function inside of `OncoKB.utils.compareIcons`. We should move `OncoKB.utils.compareIcons` into your AnnotationColumn.js. The DataTables's sorting methods are placed inside of OncoKBConnector, but they are not coupled, you can move them into AnnotationColumn.js if you need to.

**Comment by onursumer** [comment]
Thanks for the clarification Hongxin! We can also consider just modifying the hotspot category comparison function instead of adding another category. 3d hotspots are still hotspots :)

@jjgao can you send an example query where 3d hotspots don't work?

**Comment by jjgao** [comment]
@onursumer here is one example: https://cbioportal-pr-2052.herokuapp.com/cross_cancer.do?cancer_study_list=&cancer_study_id=all&data_priority=0&case_ids=&patient_case_select=sample&gene_set_choice=user-defined-list&gene_list=MAP2K1&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit&Action=Submit#crosscancer/overview/0/MAP2K1/hnc_mskcc_2016%2Csarc_mskcc%2Csarc_tcga%2Cthyroid_mskcc_2016%2Cacc_tcga%2Cchol_jhu_2013%2Cchol_nccs_2013%2Cchol_nus_2012%2Cchol_tcga%2Cgbc_shanghai_2014%2Cblca_bgi%2Cblca_dfarber_mskcc_2014%2Cblca_mskcc_solit_2012%2Cblca_mskcc_solit_2014%2Cblca_plasmacytoid_mskcc_2016%2Cblca_tcga%2Cblca_tcga_pub%2Ces_dfarber_broad_2014%2Ces_iocurie_2014%2Ccoadread_dfci_2016%2Ccoadread_genentech%2Ccoadread_mskcc%2Ccoadread_tcga%2Ccoadread_tcga_pub%2Ccesc_tcga%2Cmm_broad%2Cmds_tokyo_2011%2Cbrca_bccrc%2Cbrca_bccrc_xenograft_2014%2Cbrca_broad%2Cbrca_metabric%2Cbrca_sanger%2Cbrca_tcga%2Cbrca_tcga_pub%2Cbrca_tcga_pub2015%2Clgg_tcga%2Clgg_ucsf_2014%2Clgggbm_tcga_pub%2Cpcpg_tcga%2Cescc_icgc%2Cescc_ucla_2014%2Cegc_tmucih_2015%2Chnsc_broad%2Chnsc_jhu%2Chnsc_mdanderson_2013%2Chnsc_tcga%2Chnsc_tcga_pub%2Cnpc_nusingapore%2Cliad_inserm_fr_2014%2Clihc_amc_prv%2Clihc_riken%2Clihc_tcga%2Cnsclc_tcga_broad_2016%2Cpaac_jhu_2014%2Cpaad_icgc%2Cpaad_qcmg_uq_2016%2Cpaad_tcga%2Cpaad_utsw_2015%2Cpanet_jhu_2011%2Cpanet_shanghai_2013%2Cmeso_tcga%2Cplmeso_nyu_2015%2Cnepc_wcm_2016%2Cprad_broad%2Cprad_broad_2013%2Cprad_fhcrc%2Cprad_mich%2Cprad_mskcc%2Cprad_mskcc_2014%2Cprad_mskcc_cheny1_organoids_2014%2Cprad_su2c_2015%2Cprad_tcga%2Cprad_tcga_pub%2Ccscc_dfarber_2015%2Crms_nih_2014%2Ctgct_tcga%2Ctet_nci_2014%2Cthca_tcga%2Cthca_tcga_pub%2Cucec_tcga%2Cucec_tcga_pub%2Call_stjude_2013%2Call_stjude_2015%2Claml_tcga%2Claml_tcga_pub%2Ccll_iuopa_2015%2Clcll_broad_2013%2Cacbc_mskcc_2015%2Cnbl_amc_2012%2Cnbl_ucologne_2015%2Cmbl_broad_2012%2Cmbl_icgc%2Cmbl_pcgp%2Cesca_broad%2Cesca_tcga%2Cstad_pfizer_uhongkong%2Cstad_tcga%2Cstad_tcga_pub%2Cstad_uhongkong%2Cstad_utokyo%2Cuvm_tcga%2Cacyc_fmi_2014%2Cacyc_mda_2015%2Cacyc_mskcc_2013%2Cacyc_sanger_2013%2Cccrcc_irc_2014%2Cccrcc_utokyo_2013%2Ckirc_bgi%2Ckirc_tcga%2Ckirc_tcga_pub%2Cnccrcc_genentech_2014%2Csclc_clcgp%2Csclc_jhu%2Csclc_ucologne_2015%2Cluad_broad%2Cluad_mskcc_2015%2Cluad_tcga%2Cluad_tcga_pub%2Cluad_tsp%2Clusc_tcga%2Clusc_tcga_pub%2Ccellline_ccle_broad%2Ccellline_nci60%2Cscco_mskcc%2Cmpnst_mskcc%2Cskcm_broad%2Cskcm_broad_dfarber%2Cskcm_tcga%2Cskcm_yale%2Cdesm_broad_2015%2Cthym_tcga%2Cucs_jhu_2014%2Cucs_tcga%2Cgbm_tcga%2Cgbm_tcga_pub%2Cgbm_tcga_pub2013%2Ckirp_tcga%2Ckich_tcga%2Ckich_tcga_pub%2Curcc_mskcc_2016%2Cdlbc_broad_2012%2Cdlbc_tcga%2Cpcnsl_mayo_2015%2Cmcl_idibips_2013%2Cctcl_columbia_2015%2Cov_tcga%2Cov_tcga_pub

@onursumer: Let's fix it by modifying the hotspot category comparison function. In the new/refactored implementation of the table, this code should probably moved out of `OncoKBConnector`.

@zhx828 let's should try to make `OncoKBConnector` a js lib that does not depend on cbioportal.

**Comment by onursumer** [comment]
@jjgao Sorting should be fixed for the 3D hotspots.

**Pull request merged**